src/cargo/core/resolver/encode.rs

```diff
@@ -408,68 +416,98 @@ impl EncodableResolve {
+            versions
+                .iter()
+                .find(|package_id| {
+                    package_id.version().cmp_precedence(&pkg_version) == std::cmp::Ordering::Equal
+                })
+                .or_else(|| versions.last())
+        })
     }
-    for deps in ws.root_patch()?.values() {
-        for dep in deps {
+
+    /// Return all path dependencies recursively in given workspace
+    fn from_workspace(ws: &Workspace<'_>) -> CargoResult<Self> {
+        // If a crate is **not** a path source, then we're probably in a situation
+        // such as `cargo install` with a lock file from a remote dependency. In
+        // that case we don't need to fixup any path dependencies (as they're not
+        // actually path dependencies any more), so we ignore them.
+        let members = ws
+            .members()
+            .filter(|p| p.package_id().source_id().is_path())
+            .collect::<Vec<_>>();
+
+        let mut ret = HashSet::new();
+        let mut visited = HashSet::new();
+        for member in members.iter() {
+            ret.insert(member.package_id());
+            visited.insert(member.package_id().source_id());
+        }
+        for member in members.iter() {
+            build_pkg(member, ws, &mut ret, &mut visited);
+        }
+        for deps in ws.root_patch()?.values() {
+            for dep in deps {
+                build_dep(dep, ws, &mut ret, &mut visited);
+            }
+        }
+        for (_, dep) in ws.root_replace() {
             build_dep(dep, ws, &mut ret, &mut visited);
         }
-    }
-    for (_, dep) in ws.root_replace() {
-        build_dep(dep, ws, &mut ret, &mut visited);
-    }
 
-    return Ok(ret);
+        let path_deps = {
+            // Mapping from package name to package ids
+            let mut deps: HashMap<String, Vec<PackageId>> = HashMap::new();
+            for package_id in ret {
+                deps.entry(package_id.name().to_string())
+                    .or_default()
+                    .push(package_id);
+            }
+            deps
+        };
+
+        return Ok(Self { path_deps });
 
-    fn build_pkg(
-        pkg: &Package,
-        ws: &Workspace<'_>,
-        ret: &mut HashMap<String, SourceId>,
-        visited: &mut HashSet<SourceId>,
-    ) {
-        for dep in pkg.dependencies() {
-            build_dep(dep, ws, ret, visited);
+        fn build_pkg(
+            pkg: &Package,
+            ws: &Workspace<'_>,
+            ret: &mut HashSet<PackageId>,
+            visited: &mut HashSet<SourceId>,
+        ) {
+            for dep in pkg.dependencies() {
+                build_dep(dep, ws, ret, visited);
+            }
         }
-    }
 
-    fn build_dep(
-        dep: &Dependency,
-        ws: &Workspace<'_>,
-        ret: &mut HashMap<String, SourceId>,
-        visited: &mut HashSet<SourceId>,
-    ) {
-        let id = dep.source_id();
-        if visited.contains(&id) || !id.is_path() {
```

Comment:
`!id.is_path()` this was removed, since non-path dependencies were filtered in the beginning of `from_workspace()`

Reply:
Oops, seems to make `patch::patch_git` fail. Reverted.